src/inputs.rs

```diff
@@ -0,0 +1,209 @@
+use crate::recovery::model::{StateBytes, StateKey, StepId, StepStateBytes};
+use crate::recovery::operators::Route;
+use crate::unwrap_any;
+use pyo3::exceptions::PyTypeError;
+use pyo3::prelude::*;
+use pyo3::types::PyIterator;
+
+/// Represents a `bytewax.inputs.CustomPartInput` from Python.
+#[derive(Clone)]
+pub(crate) struct CustomPartInput(Py<PyAny>);
+
+/// Do some eager type checking.
+impl<'source> FromPyObject<'source> for CustomPartInput {
+    fn extract(ob: &'source PyAny) -> PyResult<Self> {
+        let abc = ob
+            .py()
+            .import("bytewax.inputs")?
+            .getattr("CustomPartInput")?
+            .extract()?;
+        if !ob.is_instance(abc)? {
+            Err(PyTypeError::new_err(
+                "input must derive from `bytewax.inputs.CustomPartInput`",
+            ))
+        } else {
+            Ok(Self(ob.into()))
+        }
+    }
+}
+
+impl IntoPy<Py<PyAny>> for CustomPartInput {
+    fn into_py(self, _py: Python<'_>) -> Py<PyAny> {
+        self.0
+    }
+}
+
+/// The total number of partitions in an input source.
+///
+/// Not just on this worker.
+pub(crate) struct TotalPartCount(pub(crate) usize);
+
+impl CustomPartInput {
+    /// Build all partitions for this input for this worker.
+    pub(crate) fn build_parts(
+        &self,
+        py: Python,
+        step_id: StepId,
+        index: WorkerIndex,
+        worker_count: WorkerCount,
+        mut resume_state: StepStateBytes,
+    ) -> PyResult<(PartBundle, TotalPartCount)> {
+        let mut keys: Vec<StateKey> = self.0.call_method0(py, "list_parts")?.extract(py)?;
+        keys.sort();
+
+        let part_count = TotalPartCount(keys.len());
+
+        // Can't use collect because of `?`s.
+        let mut parts = HashMap::new();
+        // We are using the [`StateKey`] routing hash as the way to
+        // divvy up partitions to workers. This is kinda an abuse of
+        // behavior, but also means we don't have to find a way to
+        // propogate the correct partition:worker mappings into the
+        // restore system, which would be more difficult as we have to
+        // find a way to treat this kind of state key differently. I
+        // might regret this.
+        for key in keys
+            .into_iter()
+            .filter(|key| key.is_local(index, worker_count))
+        {
+            let state = resume_state
+                .remove(&key)
+                .map(StateBytes::de::<TdPyAny>)
+                .unwrap_or_else(|| py.None().into());
+            tracing::info!("{index:?} building input {step_id:?} partition {key:?} with resume state {state:?}");
+            let iter: Py<PyIterator> = self
+                .0
+                .call_method1(py, "build_part", (key.clone(), state.clone_ref(py)))?
+                .as_ref(py)
+                .iter()?
+                .into();
+            let part = PartIter { iter, state };
+            parts.insert(key, part);
+        }
```

Comment:
:D Are you completely sure about that first comment?
Not saying you should, but Result [implements FromIterator](http://doc.rust-lang.org/std/result/enum.Result.html#method.from_iter), so you can move the Result outside and iterators will take care of the rest (including stopping iteration if an error is found):

```rust
        let parts = keys
            .into_iter()
            .filter(|key| key.is_local(index, worker_count))
            .map(|key| {
                let state = resume_state
                    .remove(&key)
                    .map(StateBytes::de::<TdPyAny>)
                    .unwrap_or_else(|| py.None().into());
                tracing::info!("{index:?} building input {step_id:?} partition {key:?} with resume state {state:?}");
                let iter: Py<PyIterator> = self
                    .0
                    .call_method1(py, "build_part", (key.clone(), state.clone_ref(py)))?
                    .as_ref(py)
                    .iter()?
                    .into();
                let part = PartIter { iter, state };
                Ok((key, part))
            }).collect::<PyResult<HashMap<StateKey, PartIter>>>()?;

```

Reply:
This is great.